pkg/apis/agones/v1/gameserver.go

```diff
@@ -472,151 +473,97 @@ func (gs *GameServer) applyCountsListsStatus() {
+			allErrs = append(allErrs, field.Required(fldPath.Child("container"), ErrContainerRequired))
 		}
 
 		// make sure the container value points to a valid container
 		_, _, err := gss.FindContainer(gss.Container)
 		if err != nil {
-			causes = append(causes, metav1.StatusCause{
-				Type:    metav1.CauseTypeFieldValueInvalid,
-				Field:   "container",
-				Message: err.Error(),
-			})
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("container"), gss.Container, err.Error()))
 		}
 
 		// no host port when using dynamic PortPolicy
-		for _, p := range gss.Ports {
+		for i, p := range gss.Ports {
+			path := fldPath.Child("ports").Index(i)
 			if p.PortPolicy == Dynamic || p.PortPolicy == Static {
 				if p.ContainerPort <= 0 {
-					causes = append(causes, metav1.StatusCause{
-						Type:    metav1.CauseTypeFieldValueInvalid,
-						Field:   fmt.Sprintf("%s.containerPort", p.Name),
-						Message: ErrContainerPortRequired,
-					})
+					allErrs = append(allErrs, field.Required(path.Child("containerPort"), ErrContainerPortRequired))
 				}
 			}
 
 			if p.PortPolicy == Passthrough && p.ContainerPort > 0 {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueInvalid,
-					Field:   fmt.Sprintf("%s.containerPort", p.Name),
-					Message: ErrContainerPortPassthrough,
-				})
+				allErrs = append(allErrs, field.Required(path.Child("containerPort"), ErrContainerPortPassthrough))
 			}
 
 			if p.HostPort > 0 && (p.PortPolicy == Dynamic || p.PortPolicy == Passthrough) {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueInvalid,
-					Field:   fmt.Sprintf("%s.hostPort", p.Name),
-					Message: ErrHostPort,
-				})
+				allErrs = append(allErrs, field.Forbidden(path.Child("hostPort"), ErrHostPort))
 			}
 
 			if p.Container != nil && gss.Container != "" {
 				_, _, err := gss.FindContainer(*p.Container)
 				if err != nil {
-					causes = append(causes, metav1.StatusCause{
-						Type:    metav1.CauseTypeFieldValueInvalid,
-						Field:   fmt.Sprintf("%s.container", p.Name),
-						Message: ErrContainerNameInvalid,
-					})
+					allErrs = append(allErrs, field.Invalid(path.Child("container"), *p.Container, ErrContainerNameInvalid))
 				}
 			}
 		}
-		for _, c := range gss.Template.Spec.Containers {
-			validationErrors := validateResources(c)
-			for _, err := range validationErrors {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueInvalid,
-					Field:   "container",
-					Message: err.Error(),
-				})
-			}
+		for i, c := range gss.Template.Spec.Containers {
+			path := fldPath.Child("template", "spec", "containers").Index(i)
+			allErrs = append(allErrs, validateResources(c, path)...)
 		}
-		if productCauses := apiHooks.ValidateGameServerSpec(gss); len(productCauses) > 0 {
-			causes = append(causes, productCauses...)
+		for _, cause := range apiHooks.ValidateGameServerSpec(gss) {
+			// TODO: Use field.ErrorList
```

Comment:
Curious - any reason this is a TODO without being implemented yet? Or were you waiting on some review before implementing this?

Reply:
APIHOOK is an interface, and I plan to modify the implementation of the interface in the next phase to avoid making too many changes in this PR.

If you think you can do it together in this PR, I will do the work together.

Reply:
Got it - that's fine - just wanted to check 👍🏻

Reply:
Just a workflow nit - I don't see it much in this project, but consider using something like `TODO(#xyz)` where `xyz` is an issue number. Then you can search for it easily to ensure you've accounted for all the TODOs when you're working on a long arc of PRs.